src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs

```diff
@@ -66,70 +65,183 @@ public void WriteMessage(HubMessage message, Stream output)
+                JToken resultToken = null;
+                object[] arguments = null;
+                JArray argumentToken = null;
+                ExceptionDispatchInfo argumentBindingException = null;
+                Dictionary<string, string> headers = null;
+
+                using (var reader = new JsonTextReader(textReader))
                 {
-                    // PERF: Could probably use the JsonTextReader directly for better perf and fewer allocations
-                    var token = JToken.ReadFrom(reader);
+                    reader.ArrayPool = JsonArrayPool<char>.Shared;
 
-                    if (token == null || token.Type != JTokenType.Object)
-                    {
-                        throw new InvalidDataException($"Unexpected JSON Token Type '{token?.Type}'. Expected a JSON Object.");
-                    }
+                    reader.Read();
 
-                    var json = (JObject)token;
+                    // We're always parsing a JSON object
+                    if (reader.TokenType != JsonToken.StartObject)
+                    {
+                        if (reader.TokenType == JsonToken.None)
+                        {
+                            throw new InvalidDataException("Error reading JSON.");
+                        }
 
-                    // Determine the type of the message
-                    var type = JsonUtils.GetRequiredProperty<int>(json, TypePropertyName, JTokenType.Integer);
+                        throw new InvalidDataException($"Unexpected JSON Token Type '{reader.TokenType}'. Expected a JSON Object.");
+                    }
 
-                    switch (type)
+                    do
                     {
-                        case HubProtocolConstants.InvocationMessageType:
-                            return BindInvocationMessage(json, binder);
-                        case HubProtocolConstants.StreamInvocationMessageType:
-                            return BindStreamInvocationMessage(json, binder);
-                        case HubProtocolConstants.StreamItemMessageType:
-                            return BindStreamItemMessage(json, binder);
-                        case HubProtocolConstants.CompletionMessageType:
-                            return BindCompletionMessage(json, binder);
-                        case HubProtocolConstants.CancelInvocationMessageType:
-                            return BindCancelInvocationMessage(json);
-                        case HubProtocolConstants.PingMessageType:
-                            return PingMessage.Instance;
-                        default:
-                            throw new InvalidDataException($"Unknown message type: {type}");
+                        switch (reader.TokenType)
+                        {
+                            case JsonToken.PropertyName:
+                                // We only care about parsing top level properties
+                                if (reader.Depth == 1)
```

Comment:
I'm not a fan of this. Better to `reader.Skip()` if we don't care about the content in a property. Then we know we're always at the root object here and the only two token types to worry about are properties and the end object.

Reply:
Can you clarify? What exactly would I change the code to?

Reply:
Is it basically the default case?

Reply:
Example of simply reading an object here https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Converters/ExpandoObjectConverter.cs#L113-L141

The important thing is if you don't care about a property you need to `reader.Skip()` to completely read past its value. Then the next read will either be the next property name or the end object.

The code I linked to is safer than what you're doing. Your current code wouldn't error if the object suddenly cut off with no end object, e.g.

```
{
   "type":1,
```